File: 46.全排列.java
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;

/*
 * @lc app=leetcode.cn id=46 lang=java
 *
 * [46] 全排列
 */

// @lc code=start
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        int len = nums.length;
        List<List<Integer>>res =new ArrayList<>();
        // 用栈 
        Deque<Integer> path =  new LinkedList<>();
        boolean[] used = new boolean[len];
        dfs(nums,len,0,path,used,res);
        return res;
    }
    private void dfs(int[]nums,int len,int depth,Deque<Integer> path,boolean[]used,List<List<Integer>>res){
        if(len == depth){
            // tmp只有一份，在遍历过程中会add，remove，最终回溯会root的时候，tmp就是空的。所以那里每次都要new一个对象，将tmp拷贝过去。
            res.add(new ArrayList<>(path));
            return ; 
        }
        // 定位 一个没被用的元素的下标i 
        for (int i=0;i<len;i++){
            if (used[i]){
                continue;
            }
            path.push(nums[i]);
            used[i]=true;
            dfs(nums,len,depth+1,path,used,res);
            path.pop();
            used[i]=false;
        }

    }
}
// @lc code=end

File: 46.全排列.js
~/*
 * @lc app=leetcode.cn id=46 lang=javascript
 *
 * [46] 全排列
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const res=[]
    const path=[]
    backTracking(nums,nums.length,[])
    return res
    function backTracking(n,k,used){
        if(path.length===k){
            res.push(Array.from(path))
            return 
        }
        for(let i=0;i<k;i++){
            if(used[i]) continue
            path.push(n[i])
            used[i]=true
            backTracking(n,k,used)
            path.pop()
            used[i]=false
        }
    }
};


// var permute = function (nums) {
//     let res = [];
//     let path = [];
//     let used = new Array(nums.length).fill(false);
//     dfs(nums, path, res, used, 0);
//     return res;
//   };

// let dfs = (nums, path, res, used, deep) => {
//     if (deep == nums.length) {
//       res.push(Array.from(path));
//       return;
//     }
//     for (let i = 0; i < nums.length; i++) {
//       if (!used[i]) {
//         path.push(nums[i]);
//         used[i] = true;
//         dfs(nums, path, res, used, deep + 1);
//         used[i] = false;
//         path.pop();
//       } else {
//         continue;
//       }
//     }
//   };
// @lc code=end

